llama_index/indices/knowledge_graph/retrievers.py

```diff
@@ -123,12 +136,39 @@ def _retrieve(
 
         if self._retriever_mode != KGRetrieverMode.EMBEDDING:
             for keyword in keywords:
-                cur_rel_texts = self._index_struct.get_rel_map_texts(keyword)
-                rel_texts.extend(cur_rel_texts)
-                cur_rel_map[keyword] = self._index_struct.get_rel_map_tuples(keyword)
-                if self._include_text:
-                    for node_id in self._index_struct.get_node_ids(keyword):
-                        chunk_indices_count[node_id] += 1
+                subjs = set((keyword,))
+                if self.explore_global_knowledge:
+                    # Get nodes from keyword search, and add them to the subjs set.
+                    # This helps introduce more global knowledge into the query.
+                    # While it's more expensive, thus to be turned off by default,
+                    # it can be useful for some applications.
```

Comment:
Just to confirm my understanding, this expands the search of the kg by including the neighboring relations? Or how does this differ from `graph_store_query_depth` ?

Reply:
Great question Logan, they are somewhat similar indeed, while this option enables more knowledge to be involved as it adds node-level locality: the kg(with n depth) traverse from subj not just question keywords(say x), but also from subjs in nodes with question-keywords, i.e. node 96 with x, so `x -> k` , `a -> d` can be added.

Here `a -> d` is actually already added when depth >= 2, but the knowledge of `x -> k` is only added when this option enabled.

<img width="540" alt="llama_index_global_explore" src="https://github.com/jerryjliu/llama_index/assets/1651790/fb3cb5fd-a661-4521-bc4b-4a100258af01">


```
┌──────────────────┬──────────────────┬──────────────────┬──────────────────┐
│ .─.       .─.    │  .─.       .─.   │            .─.   │  .─.       .─.   │
│( x )─────▶ y )   │ ( x )─────▶ a )  │           ( j )  │ ( m )◀────( x )  │
│ `▲'       `─'    │  `─'       `─'   │            `─'   │  `─'       `─'   │
│  │     1         │        2         │        3    │    │        4         │
│ .─.              │                  │            .▼.   │                  │
│( z )─────────────┼──────────────────┼──────────▶( i )─┐│                  │
│ `◀────┐          │                  │            `─'  ││                  │
├───────┼──────────┴──────────────────┴─────────────────┼┴──────────────────┤
│       │                      Docs/Knowledge           │                   │
│       │                            ...                │                   │
│       │                                               │                   │
├───────┼──────────┬──────────────────┬─────────────────┼┬──────────────────┤
│  .─.  └──────.   │  .─.        .─.  │                 ││  .─.             │
│ ( x )─────( b )  │ ( a )      ( x ) │                 └┼▶( n )            │
│  `─'       `─'   │  `─'        `─'  │                  │  `─'             │
│        95   │    │   │    96    │   │                  │   │              │
│            .▼.   │  .▼.        .▼.  │                  │   ▼              │
│           ( c )  │ ( d )      ( k ) │                  │  .─.             │
│            `─'   │  `─'        `─'  │                  │ ( x )            │
└──────────────────┴──────────────────┴──────────────────┴──`─'─────────────┘
```

Do you think `explore_global_knowledge` is properly-named and worthy to be added here?

Thanks!

Reply:
In the future, I am thinking of adding vector index/search in NebulaGraph, in that case, the node(doc chunk) could be persisted as another type of vertex in NebulaGraph(with its embeddings) and we could do topk distance search on the graph between vertices(for both subj, obj and doc chunks), in that case, if we wire the chunks--> subjs relationships in NebulaGrpah, then the wo could somewhat be almost the same/much similar as your understanding from intuition.

I am still studying this(all-in-one graph+doc+vector storage) now, what do you think, please?